lib/iris/experimental/ugrid.py

```diff
@@ -2115,6 +2116,252 @@ def face_node(self):
         return self._members["face_node_connectivity"]
 
 
+class MeshCoord(AuxCoord):
+    """
+    Geographic coordinate values of data on an unstructured mesh.
+
+    A MeshCoord contains a `~iris.experiomental.ugrid.Mesh`.
+    When contained in a `~iris.cube.Cube` it connects the cube to the Mesh,
+    and records (a) which cube dimension is mapped to the mesh, and (b) on
+    which mesh location (e.g. 'face' or 'node') the cube data exists.
+
+    A MeshCoord also specifies its 'axis' : 'x' or 'y'.  Its values are then,
+    accordingly, longitudes or latitudes.  The values are taken from the
+    appropriate element coordinates and connectivities in the Mesh, determined
+    by its 'location' and 'axis'.
+
+    Any cube with data on a mesh will have a MeshCoord for each axis,
+    i.e. an 'X' and a 'Y'.
+
+    The points and bounds contain coordinate values for the mesh elements,
+    which depends on location.
+    For 'node', the ".points" contains node locations.
+    For 'edge', the ".bounds" contains edge endpoints, and the points contain
+    edge locations (typically centres), if the Mesh contains them (optional).
+    For 'face', the bounds contain the face corners, and the points contain the
+    face location (optionally).
+
+    .. note::
+        As described above, it is possible for a MeshCoord to have bounds but
+        no points.  This is not possible for a regular AuxCoord or DimCoord.
+
+    .. note::
+        A MeshCoord can not yet actually be created with bounds but no points.
+        This is intended in future, but for now it raises an error.
+
+    """
+
+    # Note: effectively this list of valid mesh locations is determined by
+    # the Mesh api, but it doesn't expose those choices in a generic form, so
+    # we will just define our own here.
```

Comment:
Could add a `# TODO:` comment? I've been doing that with my data model work so far.

Reply:
ok